EventFilter/SiPixelRawToDigi/plugins/SiPixelDigiToRaw.cc

```diff
@@ -16,136 +30,142 @@
+  explicit SiPixelDigiToRaw( const edm::ParameterSet& );
+
+
+  /// get data, convert to raw event, attach again to Event
+  void produce( edm::StreamID, edm::Event&, const edm::EventSetup& ) const final;
+
+  std::shared_ptr<pr::Cache> globalBeginLuminosityBlock(edm::LuminosityBlock const&, 
+                                                        edm::EventSetup const& iES) const final;
+
+  void globalEndLuminosityBlock(edm::LuminosityBlock const&,
+                                edm::EventSetup const& iES) const final {}
+  
+  // Fill parameters descriptions
+  static void fillDescriptions(edm::ConfigurationDescriptions & descriptions);
+
+private:
+
+  mutable std::atomic_flag lock_{ ATOMIC_FLAG_INIT };
+  CMS_THREAD_GUARD(lock_) mutable edm::ESWatcher<SiPixelFedCablingMapRcd> recordWatcher;
+  CMS_THREAD_GUARD(lock_) mutable std::shared_ptr<pr::Cache> previousCache_;
+  const edm::EDGetTokenT<edm::DetSetVector<PixelDigi>> tPixelDigi; 
+  const edm::EDPutTokenT<FEDRawDataCollection> putToken_;
+  const bool usePilotBlade = false;  // I am not yet sure we need it here?
+  const bool usePhase1;
+};
 
 using namespace std;
 
 SiPixelDigiToRaw::SiPixelDigiToRaw( const edm::ParameterSet& pset ) :
-  frameReverter_(nullptr),
-  config_(pset),
-  hCPU(nullptr), hDigi(nullptr)
+  tPixelDigi{ consumes<edm::DetSetVector<PixelDigi> >(pset.getParameter<edm::InputTag>("InputLabel")) },
+  putToken_{produces<FEDRawDataCollection>()},
+  usePhase1{ pset.getParameter<bool> ("UsePhase1") }
 {
 
-  tPixelDigi = consumes<edm::DetSetVector<PixelDigi> >(config_.getParameter<edm::InputTag>("InputLabel")); 
 
   // Define EDProduct type
-  produces<FEDRawDataCollection>();
-
-  // start the counters
-  eventCounter = 0;
-  allDigiCounter = 0;
-  allWordCounter = 0;
-
-  // Timing
-  bool timing = config_.getUntrackedParameter<bool>("Timing",false);
-  if (timing) {
-    theTimer.reset(new edm::CPUTimer); 
-    hCPU = new TH1D ("hCPU","hCPU",100,0.,0.050);
-    hDigi = new TH1D("hDigi","hDigi",50,0.,15000.);
-  }
 
-  // Control the usage of phase1
-  usePhase1 = false;
-  if (config_.exists("UsePhase1")) {
-    usePhase1 = config_.getParameter<bool> ("UsePhase1");
-    if(usePhase1) edm::LogInfo("SiPixelRawToDigi")  << " Use pilot blade data (FED 40)";
-  }
+  if(usePhase1) edm::LogInfo("SiPixelRawToDigi")  << " Use pilot blade data (FED 40)";
 
-  usePilotBlade=false; // I am not yet sure we need it here?
 }
 
 // -----------------------------------------------------------------------------
-SiPixelDigiToRaw::~SiPixelDigiToRaw() {
-  delete frameReverter_;
+std::shared_ptr<pr::Cache> 
+SiPixelDigiToRaw::globalBeginLuminosityBlock(edm::LuminosityBlock const&, 
+                                             edm::EventSetup const& es) const {
+  while(lock_.test_and_set(std::memory_order_acquire)); //spin
+  auto rel = [](std::atomic_flag* f) { f->clear(std::memory_order_release); };
+  std::unique_ptr<std::atomic_flag, decltype(rel)> guard(&lock_, rel);
```

Comment:
I wish there were a simpler version compared to these 3 lines.

Reply:
If you prefer, we could switch to a boost spinlock
https://www.boost.org/doc/libs/1_54_0/doc/html/atomic/usage_examples.html#boost_atomic.usage_examples.example_spinlock.usage

but that would add a dependency on boost.